tc/core/tensor.h

```diff
@@ -0,0 +1,156 @@
+
+// Return non-owning raw pointers to DLTensor from metadata-owning DLTensorUPtr
+std::vector<const DLConstTensor*> extractRawPtrs(
+    const std::vector<DLConstTensorUPtr>& uptrs);
+std::vector<const DLTensor*> extractRawPtrs(
+    const std::vector<DLTensorUPtr>& uptrs);
+
+/**
+ * TensorInfo wraps the necessary tensor information to compile TCs and
+ * interact with the various caches.
+ * Notably, it contains alignment information but no underlying data pointer.
+ * It is serializable to protobuf and stored directly in the cache.
+ */
+struct TensorInfo {
+  DLDataType dtype;
+  uint64_t alignment;
+  std::vector<int64_t> shape;
+  std::vector<int64_t> strides;
+
+  TensorInfo(
+      DLDataType dtype,
+      uint64_t alignment,
+      const std::vector<int64_t>& shape,
+      const std::vector<int64_t>& strides);
+  explicit TensorInfo(const DLTensor* t);
+  explicit TensorInfo(const DLConstTensor* t);
+  explicit TensorInfo(const TensorInfoProto& buf);
+
+  bool operator==(const TensorInfo& t) const;
+  bool operator<(const TensorInfo& t) const;
+  TensorInfoProto toProtobuf() const;
+};
+
+template <typename T>
+std::vector<T> makeStridesFromSizes(const std::vector<T>& sizes);
+
+// Specializes for DLTensor, DLConstTensor
+template <typename DLTensorPtrType>
+std::vector<TensorInfo> makeTensorInfoVector(
+    const std::vector<DLTensorPtrType>& ts);
+std::vector<TensorInfo> makeTensorInfoVector(
+    const google::protobuf::RepeatedPtrField<TensorInfoProto>& buf);
+
+// Basic support functions for DLTensors
+DLContext getCPUDLContext();
+DLContext getGPUDLContext(int device_id = 0);
+// Specialize to various types (float, int, ...) and returns the DLDataType
+// representation
+template <typename T>
+DLDataType getDLDataType();
+bool operator==(const DLDataType& t1, const DLDataType& t2);
+
+// Print the metadata for DLDataType and DLTensor
+std::string toString(const DLDataType& t);
+std::ostream& operator<<(std::ostream& os, const DLDataType& t);
+std::ostream& operator<<(std::ostream& os, const DLTensor& t);
+
+// Basic metadata-owning DLTensor, only copies the underlying raw pointer.
+DLTensorUPtr makeDLTensor(const DLTensor* ptr);
+template <typename DLTensorType>
+DLConstTensorUPtr makeDLConstTensor(const DLTensorType* ptr);
+template <typename T>
+inline DLTensorUPtr makeDLTensor(
+    DLContext ctx,
+    DLDataType dtype,
+    const std::vector<T>& sizes,
+    const std::vector<T>& strides = std::vector<T>(),
+    const void* data = nullptr,
+    uint64_t byte_offset = 0);
+template <typename T>
+inline DLConstTensorUPtr makeDLConstTensor(
+    DLContext ctx,
+    DLDataType dtype,
+    const std::vector<T>& sizes,
+    const std::vector<T>& strides = std::vector<T>(),
+    const void* data = nullptr,
+    uint64_t byte_offset = 0);
+// A metadata-owning DLTensor reconstructed from TensorInfo does not have a
+// meaningful pointer or DLContext: pointer is  nullptr and ctx is kDLCPU.
```

Comment:
nit: extra space before "nullptr"

Also, if there is no meaningful DLContext, then why not set it to zero instead of `kDLCPU`?